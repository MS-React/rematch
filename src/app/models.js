export const game = {
  state: {
    score: 0,
    success: 0,
    fails: 0,
    maxFailures: 5,
    totalToWin: 10,
    timePerProof: 30,
    playing: false,
    player: {
      name: null
    }
  },
  reducers: {
    incrementScore(state, payload) {
      return state.score + payload;
    },
    incrementSuccess(state) {
      state.success + 1;
    },
    incrementFails(state) {
      state.fails + 1;
    },
    proofsLeft(state) {
      state.totalToWin - 1;
    },
    setPlayer(state, payload) {
      state.name = payload;
    },
    pause(state) {
      state.playing = !state.playing;
    }
  }
};
