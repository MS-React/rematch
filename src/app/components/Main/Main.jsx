import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import SelectPlayer from '../SelectPlayer';
import Game from '../Game';

import '../../assets/styles/global.scss';

const Main = ({ game }) => {
  return (
    <main className="game-container">
      {
        game.player.name === null
        && <SelectPlayer />
        || <Game />
      }
    </main>
  )
};

const mapStateToProps = state => ({
  game: state.game
});

export default connect(mapStateToProps)(Main);
