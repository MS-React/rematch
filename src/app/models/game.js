export default {
  state: {
    score: 0,
    success: 0,
    fails: 0,
    maxFailures: 5,
    totalProofs: 10,
    timePerProof: 30,
    playing: false,
    started: false,
    player: {
      name: null,
    },
    proofs: [],
  },
  reducers: {
    incrementScore: (state, payload) => ({
      ...state,
      score: state.score + Number(payload),
    }),
    incrementSuccess: state => ({
      ...state,
      success: state.success + 1,
    }),
    incrementFails: state => ({
      ...state,
      fails: state.fails + 1,
    }),
    proofsLeft: state => ({
      ...state,
      totalProofs: state.totalProofs - 1,
    }),
    setPlayer: (state, payload) => ({
      ...state,
      player: {
        name: String(payload),
      },
    }),
    resumeAndPause: state => ({
      ...state,
      playing: !state.playing,
    }),
    resetGameState: state => ({
      ...state,
      score: 0,
      totalProofs: 10,
      success: 0,
      fails: 0,
    }),
    startGame: state => ({
      ...state,
      playing: true,
      started: true,
    }),
  },
};
