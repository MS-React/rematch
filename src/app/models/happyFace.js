export default {
  state: {
    display: false,
    showHappyFace: false,
    showUnhappyFace: false,
  },
  reducers: {
    showHappyFace: state => ({
      ...state,
      display: true,
      showHappyFace: true,
      showUnhappyFace: false,
    }),
    showUnhappyFace: state => ({
      ...state,
      display: true,
      showUnhappyFace: true,
      showHappyFace: false,
    }),
  },
};
