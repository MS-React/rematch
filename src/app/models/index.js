export { default as game } from './game';
export { default as happyFace } from './happyFace';
