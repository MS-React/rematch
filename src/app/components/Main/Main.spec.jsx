import React from 'react';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import { init } from '@rematch/core';
import * as models from '../../models';

import Main from './Main';

jest.mock('../../utils/proofGenerator', () => ({
  createProof: jest.fn().mockReturnValue({
    equation: '1 + 1',
    result: 2,
  }),
}));

describe('Main', () => {
  let wrapper;
  let store;

  beforeAll(() => {
    store = init({ models });
  });

  describe('WHEN game did not start yet', () => {
    beforeEach(() => {
      wrapper = mount(
        <Provider store={store}>
          <Main />
        </Provider>,
      );
    });

    it('should show the a <PlayerSelect /> and the <Game /> start button', () => {
      expect(renderer.create(wrapper.html()).toJSON())
        .toMatchSnapshot();
    });
  });

  describe('WHEN game started', () => {
    beforeEach(() => {
      store.dispatch.game.startGame();
      wrapper = mount(
        <Provider store={store}>
          <Main />
        </Provider>,
      );
    });

    it('should display the actual game and no <PlayerSelect />', () => {
      expect(renderer.create(wrapper.html()).toJSON())
        .toMatchSnapshot();
    });
  });
});
