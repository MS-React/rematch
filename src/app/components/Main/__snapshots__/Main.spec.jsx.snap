// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main WHEN game did not start yet should show the a <PlayerSelect /> and the <Game /> start button 1`] = `"<main class=\\"game-container\\"><div class=\\"selectPlayer\\"><div>Are you good with Maths? Lets see!</div><div>Choose a Name to start!</div><input type=\\"text\\"></div><button type=\\"submit\\">Play!</button></main>"`;

exports[`Main WHEN game started should display the actual game and no <PlayerSelect /> 1`] = `"<main class=\\"game-container\\"><div class=\\"game\\"><div class=\\"game-info\\"><div class=\\"playerInfo\\">Hello! </div><div class=\\"timer\\">30</div><div class=\\"score-summary\\">Total Score: 0 <br>Success: 0 <br>Failures: 0<br></div></div><div class=\\"game-proof\\">1 + 1 = ?</div><div class=\\"game-actions\\"><input type=\\"text\\"><button type=\\"submit\\" class=\\"action-confirmProof\\">Confirm</button><button type=\\"submit\\">Pause</button></div></div></main>"`;
